statemachine/statemachine.go

```diff
@@ -371,18 +371,20 @@ func (s *stateMachine) Run() (done bool) {
 
 		metafora.Infof("task=%q transitioning %s --> %s --> %s", tid, state, msg, newstate)
 
-		// Save state
-		if err := s.ss.Store(s.task, newstate); err != nil {
-			// After upgrading to 1.25.5-gke.2000 we started experiencing the metadata server throwing POD_FINDER_IP_MISMATCH
-			// errors resulting in failures authenticating to spanner. This panic will cause the pod to cyle
-			// See https://github.com/lytics/lio/issues/30414
-			if strings.Contains(err.Error(), "spanner: code = \"Unauthenticated\"") {
-				metafora.Errorf("task=%q Unable to persist state=%q due to failure to authenticate to spanner.", tid, newstate.Code)
-				panic(err)
-			}
+		// Save state - second part of logic probably should never happen
+		if msg.Code != Release || (msg.Code == Release && state.Code != newstate.Code) {
+			if err := s.ss.Store(s.task, newstate); err != nil {
+				// After upgrading to 1.25.5-gke.2000 we started experiencing the metadata server throwing POD_FINDER_IP_MISMATCH
+				// errors resulting in failures authenticating to spanner. This panic will cause the pod to cyle
+				// See https://github.com/lytics/lio/issues/30414
+				if strings.Contains(err.Error(), "spanner: code = \"Unauthenticated\"") {
+					metafora.Errorf("task=%q Unable to persist state=%q due to failure to authenticate to spanner.", tid, newstate.Code)
+					panic(err)
```

Comment:
[q] Does this have to be called in a separate goroutine? IIRC the caller of `Run()` may capture/log panics.

Reply:
It looks like the panic recovery is in little run() which is called by exec above. So think this panic is ok here.